cluster/store/store.go

```diff
@@ -638,7 +640,29 @@ func (st *Store) Apply(l *raft.Log) interface{} {
 		panic("error proto un-marshalling log data")
 	}
 
+	// schemaOnly is necessary so that on restart when we are re-applying RAFT log entries to our in-memory schema we
+	// don't update the database. This can lead to dataloss for example if we drop then re-add a class.
+	// If we don't have any last applied index on start, schema only is always false.
+	schemaOnly := st.lastAppliedIndexOnStart != 0 && l.Index <= st.lastAppliedIndexOnStart
 	defer func() {
+		// If we have an applied index from the previous store (i.e from disk). Then reload the DB once we catch up as
+		// that means we're done doing schema only.
+		if st.lastAppliedIndexOnStart != 0 && l.Index == st.lastAppliedIndexOnStart {
+			st.log.WithFields(logrus.Fields{
+				"log_type":                     l.Type,
+				"log_name":                     l.Type.String(),
+				"log_index":                    l.Index,
+				"last_store_log_applied_index": st.lastAppliedIndexOnStart,
+			}).Debug("reloading local DB as RAFT and local DB are now caught up")
+			cs := make([]command.UpdateClassRequest, len(st.db.Schema.Classes))
+			i := 0
+			for _, v := range st.db.Schema.Classes {
+				cs[i] = command.UpdateClassRequest{Class: &v.Class, State: &v.Sharding}
+				i++
+			}
```

Comment:
**thought:** is it necessary ? what I am thinking is that will cause reload to the db with every apply and could be expensive op

Reply:
This will trigger an apply only on the log entry where we have caught up as lastAppliedIndexOnStart should not be updated.

Reply:
After checking: The DB on `Open` under the hood calls `db.init` which (from the schema itself) will load indexes for all the classes. So that means that if we "always" open the DB then we need to reload here once we have caught up so that the DB opens the index for all the classes that are present.